patches/0013-p256-armv8-optimizations

```diff
@@ -2429,20 +2457,446 @@ Index: aws-lc/third_party/boringssl/crypto/fipsmodule/ec/p256-x86_64_test.cc
++    // with the following differences:
++    // - "$shift" is set directly to the number of trailing 0s in B
++    //   (using rbit and clz instructions)
++    // - The loop is only used to call SHIFT1(X)
++    //   and $shift is decreased while executing the X loop.
++    // - SHIFT256(B, $shift) is performed before right-shifting X; they are independent
++
++.Lbeeu_shift_A_Y:
++    // Same for A and Y.
++    // Afterwards, (2) still holds.
++    // Reverse the bit order of $a0
++    // $shift := number of trailing 0s in $a0 (= number of leading 0s in $t1)
++    rbit    $t1, $a0
++    clz     $shift, $t1
++
++    // If there is no shift, goto |B-A|, X+Y update
++    cbz     $shift, .Lbeeu_update_B_X_or_A_Y
++
++    // Shift A right by "$shift" bits
++    ${\SHIFT256($a0, $a1, $a2, $a3)}
++
++    // Shift Y right by "$shift" bits, adding n whenever Y becomes odd.
++    // $shift--;
++    // $t0 := 0; needed in the addition to the most significant word in SHIFT1
++    eor     $t0, $t0, $t0
++.Lbeeu_shift_loop_Y:
++    ${\SHIFT1($y0, $y1, $y2, $y3, $y4)}
++    subs    $shift, $shift, #1
++    bne     .Lbeeu_shift_loop_Y
++
++.Lbeeu_update_B_X_or_A_Y:
++    // Try T := B - A; if cs, continue with B > A (cs: carry set = no borrow)
++    subs    $t0, $b0, $a0
++    sbcs    $t1, $b1, $a1
++    sbcs    $t2, $b2, $a2
++    sbcs    $t3, $b3, $a3
++    bcs     .Lbeeu_B_greater_than_A
++
++    // Else A > B =>
++    // A := A - B; Y := Y + X; goto beginning of the loop
++    subs    $a0, $a0, $b0
++    sbcs    $a1, $a1, $b1
++    sbcs    $a2, $a2, $b2
++    sbcs    $a3, $a3, $b3
++
++    adds    $y0, $y0, $x0
++    adcs    $y1, $y1, $x1
++    adcs    $y2, $y2, $x2
++    adcs    $y3, $y3, $x3
++    adc     $y4, $y4, $x4
++    b       .Lbeeu_loop
++
++.Lbeeu_B_greater_than_A:
++    // Continue with B > A =>
++    // B := B - A; X := X + Y; goto beginning of the loop
++    mov     $b0, $t0
++    mov     $b1, $t1
++    mov     $b2, $t2
++    mov     $b3, $t3
++
++    adds    $x0, $x0, $y0
++    adcs    $x1, $x1, $y1
++    adcs    $x2, $x2, $y2
++    adcs    $x3, $x3, $y3
++    adc     $x4, $x4, $y4
++    b       .Lbeeu_loop
++
++.Lbeeu_loop_end:
++    // The Euclid's algorithm loop ends when A == gcd(a,n);
++    // this would be 1, when a and n are co-prime (i.e. do not have a common factor).
++    // Since (-1)*Y*a == A (mod |n|), Y>0
++    // then out = -Y mod n
++
++    // Verify that A = 1 ==> (-1)*Y*a = A = 1  (mod |n|)
++    // Is A-1 == 0?
++    // If not, fail.
++    sub     $t0, $a0, #1
++    orr     $t0, $t0, $a1
++    orr     $t0, $t0, $a2
++    orr     $t0, $t0, $a3
++    cbnz    $t0, .Lbeeu_err
++
++    // If Y>n ==> Y:=Y-n
++.Lbeeu_reduction_loop:
++    // x_i := y_i - n_i (X is no longer needed, use it as temp)
++    // ($t0 = 0 from above)
++    subs    $x0, $y0, $n0
++    sbcs    $x1, $y1, $n1
++    sbcs    $x2, $y2, $n2
++    sbcs    $x3, $y3, $n3
++    sbcs    $x4, $y4, $t0
```

Comment:
There is no fancy way to skip this when no borrow?

Reply:
If I understood your question correctly, I think you wonder if you can compare quickly instead of going through a full multi-precision subtraction. 
I thought about that and found that comparing will need to compare corresponding words starting from the most significant, but there are cases, you would go down to less significant words if the most significant ones are equal (e.g., zeros), then you need to select the words of the larger value to be the minuend and the other one the subtrahend. So in terms of code size, it may be larger, and the effort may not be saved, since the number of instructions may increase.
In this architecture, the pipeline can execute more instructions per clock cycle, and in the worst case, if gated on the carry, it will be one instruction per cycle.

Reply:
In the last `sbcs` instruction the subtrahend is `t0 = 0`. My assumption was that you do this to take any borrow into account. NP

Reply:
Correct.